src/event_loop_dispatcher.hpp

```diff
@@ -0,0 +1,134 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2019 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_DART_EVENT_LOOP_DISPATCHER_HPP
+#define REALM_DART_EVENT_LOOP_DISPATCHER_HPP
+
+#include <realm/object-store/util/scheduler.hpp>
+
+#include <tuple>
+#include <iostream>
+#include "dart_api_dl.h"
+
+namespace realm::util {
+
+template <class F>
+class EventLoopDispatcher;
+
+template <typename... Args>
+class EventLoopDispatcher<void(Args...)> {
+    using Tuple = std::tuple<typename std::remove_reference<Args>::type...>;
+
+private:
+    const std::shared_ptr<util::UniqueFunction<void(Args...)>> m_func;
+    std::shared_ptr<util::Scheduler> m_scheduler; //  = util::Scheduler::make_default();
+
+public:
+    EventLoopDispatcher(std::shared_ptr<util::Scheduler> scheduler, util::UniqueFunction<void(Args...)> func)
+        : m_func(std::make_shared<util::UniqueFunction<void(Args...)>>(std::move(func))), m_scheduler(scheduler)
+    {
+    }
+
+    const util::UniqueFunction<void(Args...)>& func() const
+    {
+        return *m_func;
+    }
+
+    void operator()(Args... args)
+    {
+        m_scheduler->invoke(
+            [scheduler = m_scheduler, func = m_func, args = std::make_tuple(std::forward<Args>(args)...)]() mutable {
+                std::apply(*func, std::move(args));
+                // Each invocation block will retain the scheduler, so the scheduler
```

Comment:
What happens if the user drops the Dart instance. For example tears down the isolate.

Reply:
Same misery as without the `EventLoopDispatcher`. It is the callers responsibility that the userdata is still alive when a/the callback happens. If you use a persistent handle, then the known issue with isolate shutdown in 2.16 still applies.